mongo.go

```diff
@@ -0,0 +1,472 @@
+	if err := sr.Decode(&kv); err != nil {
+		return nil, fmt.Errorf("decoding key-value: %s", err)
+	}
+	return kv.Value, nil
+}
+
+func (m *MongoDS) delete(ctx context.Context, key datastore.Key) error {
+	_, err := m.col.DeleteOne(ctx, bson.M{"_id": key.String()})
+	if err != nil {
+		return fmt.Errorf("delete document: %s", err)
+	}
+
+	return nil
+}
+
+func (m *MongoDS) put(ctx context.Context, key datastore.Key, val []byte) error {
+	_, err := m.col.UpdateOne(ctx, bson.M{"_id": key.String()}, bson.M{"$set": bson.M{"v": val}}, options.Update().SetUpsert(true))
+	if err != nil {
+		return fmt.Errorf("inserting/updating key-value: %s", err)
+	}
+	return nil
+}
+
+func (m *MongoDS) has(ctx context.Context, key datastore.Key) (bool, error) {
+	sr := m.col.FindOne(ctx, bson.M{"_id": key.String()})
+	if sr.Err() == mongo.ErrNoDocuments {
+		return false, nil
+	}
+	if sr.Err() != nil {
+		return false, fmt.Errorf("finding key: %s", sr.Err())
+	}
+	return true, nil
+}
+
+func (m *MongoDS) getSize(ctx context.Context, key datastore.Key) (int, error) {
+	v, err := m.get(ctx, key)
+	if err == datastore.ErrNotFound {
+		return -1, err
+	}
+	if err != nil {
+		return 0, fmt.Errorf("getting value: %s", err)
+	}
+	return len(v), nil
+}
+
+func (m *MongoDS) query(ctx context.Context, q dsextensions.QueryExt) (query.Results, error) {
+	opts := options.Find()
+
+	// Handle ordering
+	asc := true
+	if len(q.Orders) > 0 {
+		switch q.Orders[0].(type) {
+		case dsq.OrderByKey, *dsq.OrderByKey:
+			// We order by key by default.
+		case dsq.OrderByKeyDescending, *dsq.OrderByKeyDescending:
+			// Reverse order by key
+			asc = false
+		default:
+			// Ok, we have a weird order we can't handle. Let's
+			// perform the _base_ query (prefix, filter, etc.), then
+			// handle sort/offset/limit later.
+
+			// Skip the stuff we can't apply.
+			baseQuery := q
+			baseQuery.Limit = 0
+			baseQuery.Offset = 0
+			baseQuery.Orders = nil
+
+			// perform the base query.
+			res, err := m.query(ctx, baseQuery)
+			if err != nil {
+				return nil, err
+			}
+
+			// fix the query
+			res = dsq.ResultsReplaceQuery(res, q.Query)
+
+			// Remove the parts we've already applied.
+			naiveQuery := q.Query
+			naiveQuery.Prefix = ""
+			naiveQuery.Filters = nil
+
+			// Apply the rest of the query
+			return dsq.NaiveQueryApply(naiveQuery, res), nil
+		}
+	}
+	opts.SetSort(bson.M{"_id": 1})
+	if !asc {
+		opts.SetSort(bson.M{"_id": -1})
+	}
+
+	prefix := datastore.NewKey(q.Prefix).String()
+	// Important to consider the '/' suffix to respect Prefix semantics
+	// of returning strictly child keys.
+	var filters bson.A
+	if prefix != "/" {
+		rgx := fmt.Sprintf("^%s/.*", prefix)
+		filters = append(filters, bson.M{"_id": bson.M{"$regex": primitive.Regex{Pattern: rgx}}})
+	}
+	seekPrefix := datastore.NewKey(q.SeekPrefix).String()
+	if seekPrefix != "/" {
+		op := "$gte"
+		if !asc {
+			op = "$lte"
+		}
+		filters = append(filters, bson.M{"_id": bson.M{op: seekPrefix}})
+	}
```

Comment:
If we have a `PrefixSeek`, we use `$gte` if we're in an ascending key ordering case. And `$lte` otherwise, since you should think about the seeking in reverse.